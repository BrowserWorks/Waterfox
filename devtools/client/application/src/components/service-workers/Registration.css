/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

/*
 * The current layout of a registration is
 *
 *  +------+----------------------+----------------+
 *  | Header - scope + timestamp  | Unregister_btn |
 *  +------+----------------------+----------------|
 *  | worker 1                                     |
    | worker 2                                     |
    | ...                                          |
    +----------------------------------------------+
    |                              Unregister btn  |
    +----------------------------------------------+
 */

.registration {
  line-height: 1.5;
  font-size: var(--body-10-font-size);
  display: grid;
  grid-template-columns: minmax(0, 1fr) auto;
  grid-template-rows: minmax(calc(var(--base-unit) * 6), auto) 1fr auto;
  grid-column-gap: calc(4 * var(--base-unit));
  grid-row-gap: calc(2 * var(--base-unit));
  grid-template-areas: "header          header-controls"
                       "workers         workers"
                       "footer-controls footer-controls";
}

/* vertical layout */
@media(max-width: 700px) {
  .registration__controls {
    grid-area: footer-controls;
    justify-self: end;
  }
}

/* wide layout */
@media(min-width: 701px) {
  .registration__controls {
    grid-area: header-controls;
  }
}

.registration__header {
  grid-area: header;

  display: grid;
  grid-template-areas: "icon      scope"
                       "timestamp timestamp";
  grid-template-columns: auto 1fr;
  /* make the row min height the same size of the Unregister button,
     so we avoid a shifting movement when the button is not displayed */
  grid-template-rows: minmax(calc(var(--base-unit) * 6), auto);
  grid-column-gap: calc(var(--base-unit) * 2);
  align-items: center;
}

.registration__header::before {
  display: inline-block;
  width: calc(var(--base-unit) * 4);
  height: calc(var(--base-unit) * 4);
  content: "";
  background-image: url(chrome://devtools/skin/images/debugging-workers.svg);
}

.registration__scope {
  font-size: var(--title-10-font-size);
  font-weight: var(--title-10-font-weight);
  user-select: text;
  margin: 0;

  grid-area: scope;
}

.registration__updated-time {
  color: var(--theme-text-color-alt);
  grid-area: timestamp;
}

.registration__workers {
  grid-area: workers;
  list-style-type: none;
  padding: 0;
}

.registration__workers-item:not(:first-child) {
  margin-block-start: calc(var(--base-unit) * 2);
}
