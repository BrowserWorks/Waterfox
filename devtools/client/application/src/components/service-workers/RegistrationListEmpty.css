/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.app-page__title .external-link {
  font-weight: var(--title-30-font-weight);
}

.registration-list-empty__tips {
  margin-inline-start: 2rem;
}

.registration-list-empty__tips-item {
  margin-block-start: 0.5rem;
  margin-block-end: 0.5rem;
}
