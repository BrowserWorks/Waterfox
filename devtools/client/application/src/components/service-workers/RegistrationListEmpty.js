/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { openDocLink, openTrustedLink } = require("devtools/client/shared/link");
const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");
const {
  a,
  article,
  h1,
  li,
  p,
  ul,
} = require("devtools/client/shared/vendor/react-dom-factories");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);

const {
  services,
} = require("devtools/client/application/src/modules/application-services");

const DOC_URL =
  "https://developer.mozilla.org/docs/Web/API/Service_Worker_API/Using_Service_Workers" +
  "?utm_source=devtools&utm_medium=sw-panel-blank";

/**
 * This component displays help information when no service workers are found for the
 * current target.
 */
class RegistrationListEmpty extends PureComponent {
  switchToConsole() {
    services.selectTool("webconsole");
  }

  switchToDebugger() {
    services.selectTool("jsdebugger");
  }

  openAboutDebugging() {
    openTrustedLink("about:debugging#workers");
  }

  openDocumentation() {
    openDocLink(DOC_URL);
  }

  render() {
    return article(
      { className: "registration-list-empty js-registration-list-empty" },
      Localized(
        {
          id: "serviceworker-empty-intro",
          a: a({
            className: "external-link",
            onClick: () => this.openDocumentation(),
          }),
        },
        h1({ className: "app-page__title" })
      ),
      Localized({ id: "serviceworker-empty-suggestions" }, p({})),
      ul(
        { className: "registration-list-empty__tips" },
        Localized(
          {
            id: "serviceworker-empty-suggestions-console",
            a: a({ className: "link", onClick: () => this.switchToConsole() }),
          },
          li({ className: "registration-list-empty__tips-item" })
        ),
        Localized(
          {
            id: "serviceworker-empty-suggestions-debugger",
            a: a({ className: "link", onClick: () => this.switchToDebugger() }),
          },
          li({ className: "registration-list-empty__tips-item" })
        ),
        Localized(
          {
            id: "serviceworker-empty-suggestions-aboutdebugging",
            a: a({
              className: "link js-trusted-link",
              onClick: () => this.openAboutDebugging(),
            }),
          },
          li({ className: "registration-list-empty__tips-item" })
        )
      )
    );
  }
}

// Exports
module.exports = RegistrationListEmpty;
