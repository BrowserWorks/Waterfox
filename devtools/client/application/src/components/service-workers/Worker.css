/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

 /*
 * The current layout of a service worker item is
 *
 *  +---+----------+-------------+----------------|
 *  |     "Source" | script_name debug_link       |
 *  |--------------+-------------+----------------|
 *  |     "Status" | status start_button          |
 *  +---+----------+-------------+----------------|
 */

.worker {
  line-height: 1.5;
  font-size: var(--body-10-font-size);
}

.worker__source-url {
  user-select: text;
}

.worker__meta-name {
  color: var(--theme-text-color-alt);
}

.worker__data {
  display: grid;
  grid-template-columns: auto 1fr;
  grid-column-gap: calc(var(--base-unit) * 2);
  grid-row-gap: calc(var(--base-unit) * 0.5);
}

.worker__data > * {
  margin: 0;
}

/* skip the first <dt> + <dd> pair elements */
.worker__data > *:nth-child(n + 3) {
  margin-block-start: calc(var(--base-unit) * 2);
}

.worker__link-start,
.worker__link-debug {
  margin: 0 calc(var(--base-unit) * 2);
}

.worker__status {
  text-transform: capitalize;
  --status-bg-color: transparent;
  --status-border-color: transparent;
}

.worker__status::before {
  content: "";
  margin-inline-end: var(--base-unit);
  width: calc(var(--base-unit) * 2);
  height: calc(var(--base-unit) * 2);
  display: inline-block;
  background: var(--status-bg-color);
  border: 1px solid var(--status-border-color);
  border-radius: 100%;
}

.worker__status--active {
  --status-bg-color: var(--green-60);
  --status-border-color: var(--green-60);
}

.worker__status--waiting {
  --status-bg-color: var(--theme-text-color-alt);
  --status-border-color: var(--theme-text-color-alt);
}

.worker__status--installing, .worker__status--default {
  --status-bg-color: transparent;
  --status-border-color: var(--theme-text-color-alt);
}
