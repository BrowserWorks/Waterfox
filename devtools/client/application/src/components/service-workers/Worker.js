/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

"use strict";

const { Ci } = require("chrome");

const {
  createFactory,
  PureComponent,
} = require("devtools/client/shared/vendor/react");

const PropTypes = require("devtools/client/shared/vendor/react-prop-types");

const {
  dd,
  dl,
  dt,
  section,
  span,
} = require("devtools/client/shared/vendor/react-dom-factories");

const { getUnicodeUrlPath } = require("devtools/client/shared/unicode-url");

const FluentReact = require("devtools/client/shared/vendor/fluent-react");
const Localized = createFactory(FluentReact.Localized);
const { l10n } = require("devtools/client/application/src/modules/l10n");

const {
  services,
} = require("devtools/client/application/src/modules/application-services");
const Types = require("devtools/client/application/src/types/index");

const UIButton = createFactory(
  require("devtools/client/application/src/components/ui/UIButton")
);

/**
 * This component is dedicated to display a worker, more accurately a service worker, in
 * the list of workers displayed in the application panel. It displays information about
 * the worker as well as action links and buttons to interact with the worker (e.g. debug,
 * unregister, update etc...).
 */
class Worker extends PureComponent {
  static get propTypes() {
    return {
      isDebugEnabled: PropTypes.bool.isRequired,
      worker: PropTypes.shape(Types.worker).isRequired,
    };
  }

  constructor(props) {
    super(props);

    this.debug = this.debug.bind(this);
    this.start = this.start.bind(this);
  }

  debug() {
    if (!this.isRunning()) {
      console.log("Service workers cannot be debugged if they are not running");
      return;
    }

    services.openWorkerInDebugger(this.props.worker.workerTargetFront);
  }

  start() {
    if (!this.props.isDebugEnabled) {
      console.log("Service workers cannot be started in multi-e10s");
      return;
    }

    if (!this.isActive() || this.isRunning()) {
      console.log("Running or inactive service workers cannot be started");
      return;
    }

    const { registrationFront } = this.props.worker;
    registrationFront.start();
  }

  isRunning() {
    // We know the worker is running if it has a worker actor.
    return !!this.props.worker.workerTargetFront;
  }

  isActive() {
    return this.props.worker.state === Ci.nsIServiceWorkerInfo.STATE_ACTIVATED;
  }

  getLocalizedStatus() {
    if (this.isActive() && this.isRunning()) {
      return l10n.getString("serviceworker-worker-status-running");
    } else if (this.isActive()) {
      return l10n.getString("serviceworker-worker-status-stopped");
    }
    // NOTE: this is already localized by the service worker front
    // (strings are in debugger.properties)
    return this.props.worker.stateText;
  }

  getClassNameForStatus(baseClass) {
    const { state } = this.props.worker;

    switch (state) {
      case Ci.nsIServiceWorkerInfo.STATE_PARSED:
      case Ci.nsIServiceWorkerInfo.STATE_INSTALLING:
        return "worker__status--installing";
      case Ci.nsIServiceWorkerInfo.STATE_INSTALLED:
      case Ci.nsIServiceWorkerInfo.STATE_ACTIVATING:
        return "worker__status--waiting";
      case Ci.nsIServiceWorkerInfo.STATE_ACTIVATED:
        return "worker__status--active";
    }

    return "worker__status--default";
  }

  formatSource(source) {
    const parts = source.split("/");
    return getUnicodeUrlPath(parts[parts.length - 1]);
  }

  renderDebugButton() {
    const { isDebugEnabled } = this.props;

    const isDisabled = !this.isRunning() || !isDebugEnabled;

    const localizationId = isDebugEnabled
      ? "serviceworker-worker-debug"
      : "serviceworker-worker-debug-forbidden";

    return Localized(
      {
        id: localizationId,
        // The localized title is only displayed if the debug link is disabled.
        attrs: {
          title: isDisabled,
        },
      },
      UIButton({
        onClick: this.debug,
        className: `js-debug-button`,
        disabled: isDisabled,
        size: "micro",
      })
    );
  }

  renderStartButton() {
    const { isDebugEnabled } = this.props;

    // avoid rendering the button at all for workers that are either running,
    // or in a state that prevents them from starting (like waiting)
    if (this.isRunning() || !this.isActive()) {
      return null;
    }

    const isDisabled = !isDebugEnabled;
    return Localized(
      {
        id: "serviceworker-worker-start2",
        // The localized title is only displayed if the debug link is disabled.
        attrs: {
          title: !isDisabled,
        },
      },
      UIButton({
        onClick: this.start,
        className: `js-start-button`,
        disabled: isDisabled,
        size: "micro",
      })
    );
  }

  render() {
    const { worker } = this.props;
    const statusText = this.getLocalizedStatus();
    const statusClassName = this.getClassNameForStatus();

    return section(
      { className: "worker js-sw-worker" },
      dl(
        { className: "worker__data" },
        Localized(
          { id: "serviceworker-worker-source" },
          dt({ className: "worker__meta-name" })
        ),
        dd(
          {},
          span(
            {
              title: worker.url,
              className: "worker__source-url js-source-url",
            },
            this.formatSource(worker.url)
          ),
          " ",
          this.renderDebugButton()
        ),
        Localized(
          { id: "serviceworker-worker-status" },
          dt({ className: "worker__meta-name" })
        ),
        dd(
          {},
          span(
            { className: `js-worker-status worker__status ${statusClassName}` },
            statusText
          ),
          " ",
          this.renderStartButton()
        )
      )
    );
  }
}

module.exports = Worker;
