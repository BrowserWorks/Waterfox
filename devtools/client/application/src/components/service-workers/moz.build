# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

DevToolsModules(
    'Registration.css',
    'Registration.js',
    'RegistrationList.css',
    'RegistrationList.js',
    'RegistrationListEmpty.css',
    'RegistrationListEmpty.js',
    'Worker.css',
    'Worker.js',
    'WorkersPage.js',
)
