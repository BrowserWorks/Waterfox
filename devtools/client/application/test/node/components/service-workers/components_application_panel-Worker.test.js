/* Any copyright is dedicated to the Public Domain.
   http://creativecommons.org/publicdomain/zero/1.0/ */

"use strict";

// Import libs
const { shallow } = require("enzyme");
const { createFactory } = require("react");

const {
  WORKER_RUNNING,
  WORKER_STOPPED,
  WORKER_WAITING,
} = require("devtools/client/application/test/node/fixtures/data/constants");

const Worker = createFactory(
  require("devtools/client/application/src/components/service-workers/Worker")
);

describe("Worker", () => {
  it("Renders the expected snapshot for a running worker", () => {
    const wrapper = shallow(
      Worker({
        isDebugEnabled: true,
        worker: WORKER_RUNNING,
      })
    );

    // ensure proper status
    expect(wrapper.find(".js-worker-status").text()).toBe(
      "serviceworker-worker-status-running"
    );
    // check that Start button is not available
    expect(wrapper.find(".js-start-button")).toHaveLength(0);

    expect(wrapper).toMatchSnapshot();
  });

  it("Renders the expected snapshot for a stopped worker", () => {
    const wrapper = shallow(
      Worker({
        isDebugEnabled: true,
        worker: WORKER_STOPPED,
      })
    );

    // ensure proper status
    expect(wrapper.find(".js-worker-status").text()).toBe(
      "serviceworker-worker-status-stopped"
    );
    // check that Start button is not available
    expect(wrapper.find(".js-start-button")).toHaveLength(1);

    expect(wrapper).toMatchSnapshot();
  });

  it("Renders the expected snapshot for a non-active worker", () => {
    const wrapper = shallow(
      Worker({
        isDebugEnabled: true,
        worker: WORKER_WAITING,
      })
    );

    // ensure proper status
    // NOTE: since non-active status are localized directly in the front, not
    //       in the panel, we don't expect a localization ID here
    expect(wrapper.find(".js-worker-status").text()).toBe("installed");
    // check that Start button is not available
    expect(wrapper.find(".js-start-button")).toHaveLength(0);

    expect(wrapper).toMatchSnapshot();
  });

  it("Enables/disabled the debug button depending of debugging being available", () => {
    // check disabled debugging
    let wrapper = shallow(
      Worker({
        isDebugEnabled: false,
        worker: WORKER_RUNNING,
      })
    );
    expect(wrapper.find(".js-debug-button[disabled=true]")).toHaveLength(1);

    // check enabled debugging
    wrapper = shallow(
      Worker({
        isDebugEnabled: true,
        worker: WORKER_RUNNING,
      })
    );
    expect(wrapper.find(".js-debug-button[disabled=false]")).toHaveLength(1);
  });
});
