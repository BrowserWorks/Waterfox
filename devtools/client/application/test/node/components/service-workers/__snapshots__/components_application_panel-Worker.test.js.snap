// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Worker Renders the expected snapshot for a non-active worker 1`] = `
<section
  className="worker js-sw-worker"
>
  <dl
    className="worker__data"
  >
    <Localized
      id="serviceworker-worker-source"
    >
      <dt
        className="worker__meta-name"
      />
    </Localized>
    <dd>
      <span
        className="worker__source-url js-source-url"
        title="http://example.com/worker.js"
      >
        worker.js
      </span>
       
      <Localized
        attrs={
          Object {
            "title": true,
          }
        }
        id="serviceworker-worker-debug"
      >
        <UIButton
          className="js-debug-button"
          disabled={true}
          onClick={[Function]}
          size="micro"
        />
      </Localized>
    </dd>
    <Localized
      id="serviceworker-worker-status"
    >
      <dt
        className="worker__meta-name"
      />
    </Localized>
    <dd>
      <span
        className="js-worker-status worker__status worker__status--waiting"
      >
        installed
      </span>
       
    </dd>
  </dl>
</section>
`;

exports[`Worker Renders the expected snapshot for a running worker 1`] = `
<section
  className="worker js-sw-worker"
>
  <dl
    className="worker__data"
  >
    <Localized
      id="serviceworker-worker-source"
    >
      <dt
        className="worker__meta-name"
      />
    </Localized>
    <dd>
      <span
        className="worker__source-url js-source-url"
        title="http://example.com/worker.js"
      >
        worker.js
      </span>
       
      <Localized
        attrs={
          Object {
            "title": false,
          }
        }
        id="serviceworker-worker-debug"
      >
        <UIButton
          className="js-debug-button"
          disabled={false}
          onClick={[Function]}
          size="micro"
        />
      </Localized>
    </dd>
    <Localized
      id="serviceworker-worker-status"
    >
      <dt
        className="worker__meta-name"
      />
    </Localized>
    <dd>
      <span
        className="js-worker-status worker__status worker__status--active"
      >
        serviceworker-worker-status-running
      </span>
       
    </dd>
  </dl>
</section>
`;

exports[`Worker Renders the expected snapshot for a stopped worker 1`] = `
<section
  className="worker js-sw-worker"
>
  <dl
    className="worker__data"
  >
    <Localized
      id="serviceworker-worker-source"
    >
      <dt
        className="worker__meta-name"
      />
    </Localized>
    <dd>
      <span
        className="worker__source-url js-source-url"
        title="http://example.com/worker.js"
      >
        worker.js
      </span>
       
      <Localized
        attrs={
          Object {
            "title": true,
          }
        }
        id="serviceworker-worker-debug"
      >
        <UIButton
          className="js-debug-button"
          disabled={true}
          onClick={[Function]}
          size="micro"
        />
      </Localized>
    </dd>
    <Localized
      id="serviceworker-worker-status"
    >
      <dt
        className="worker__meta-name"
      />
    </Localized>
    <dd>
      <span
        className="js-worker-status worker__status worker__status--active"
      >
        serviceworker-worker-status-stopped
      </span>
       
      <Localized
        attrs={
          Object {
            "title": true,
          }
        }
        id="serviceworker-worker-start2"
      >
        <UIButton
          className="js-start-button"
          disabled={false}
          onClick={[Function]}
          size="micro"
        />
      </Localized>
    </dd>
  </dl>
</section>
`;
